test/test_memmove.c

```diff
@@ -0,0 +1,203 @@
+#define SIZE_VARIANCE 4
+#define BUF_DUMP_WIDTH 20
+
+static const int test_offsets[][2] = {
+	{ 0, 2000 },
+	{ 2000, 0 },
+	{ 100, 200 },
+	{ 200, 100 }
+};
+
+char pattern[BUF_LEN];
+char buffer[BUF_LEN] __attribute__ ((aligned (4))); 
+
+
+/* print bytes from buffer in hexadecimal form */
+void print_buffer(void *buffer, int n)
+{
+	int i;
+	for (i = 0; i < n; i++) {
+		if ((i % (BUF_DUMP_WIDTH)) == 0)
+			printf("\nbytes %d - %d\n", i, i + BUF_DUMP_WIDTH - 1);
+		printf("%02X ", *((unsigned char *) buffer + i));
+	}
+	printf("\n\n");
+}
+
+
+void fill_random(char *buffer, int len)
+{
+	int i;
+	srand(SEED);
+	for (i = 0; i < len; i++) {
+		buffer[i] = (char) rand();
+	}	
+}
+
+
+/* loop over buffer and check every byte - copied as well as non-modified ones */
+int memmove_check_result(char *buffer, char *pattern, int src_off, int dst_off, int len)
+{
+	char current_byte;
+	int i, err, incorrect_bytes = 0, in_dst;
+
+	for (i = 0; i < BUF_LEN; i++) {
+		err = 0;	
+		current_byte = *(buffer + i);
+		/* check if i is in destination buffer */	
+		in_dst = (i >= dst_off && i < dst_off + len);
+		
+		if (!in_dst) 
+			err = (current_byte != *(pattern + i));
+		else 
+			err = (current_byte != *(pattern + src_off + i - dst_off));
+			
+		if (err && verbose_test() == 2)
+			printf("Incorrect byte nr: %d value: %02X\n", i, (unsigned char) current_byte);
+
+		incorrect_bytes += err;
+	}
+	return incorrect_bytes;
+}
+
+
+void copy_buffer(char *buffer, char *pattern, int len)
+{
+	int i;
+	for (i = 0; i < len; i++) {
+		*(buffer + i) = *(pattern + i);
+	}
+}
+
+
+/*
+ * Perform memmove when src and dst are pointers inside one continous buffer.
+ * In loop add offset to ptrs in aligned buffer to simulate misaligned data
+ * Check if data was copied properly and if bytes around weren't violeted
+ */
+int memmove_check_var_off(char *buffer, char *pattern, int off_src, int off_dst,  int len)
+{
+	int result, failed = 0;
+	void *ptr_dst, *ptr_src;	
+	int off_src_shifted, off_dst_shifted, delta_off_src, delta_off_dst;
+	
+	for (delta_off_src = 0;  delta_off_src < OFFSET_VARIANCE; delta_off_src++) {
+		for (delta_off_dst = 0;  delta_off_dst < OFFSET_VARIANCE; delta_off_dst++) {
+		
+			off_src_shifted = off_src + delta_off_src;
+			off_dst_shifted = off_dst + delta_off_dst;
+			ptr_dst = buffer + off_dst_shifted;
+			ptr_src = buffer + off_src_shifted;	
+			
+			copy_buffer(buffer, pattern, BUF_LEN); 
+			memmove(ptr_dst,  ptr_src, len);
+			result = memmove_check_result(buffer, pattern, off_src_shifted, off_dst_shifted, len);
+			failed += !!result;	
+
+			/* print failed test details */
+			if (result && verbose_test()) {
+				printf("\nTestcase failed. Incorrect bytes: %d \n", result);
+				printf("len %d off_src %d delta_off_src %d off_dst %d delta_off_dst %d\n", 
+						len, off_src, delta_off_src, off_dst, delta_off_dst);
+				if (verbose_test() > 1) {
+					printf("Dumping output buffer\n");
+					print_buffer(buffer, BUF_LEN);
+				}
```

Comment:
If test should fail, many bytes will cause error. It will cause rain of erros msg and buffer dumps. I think you should end test after buffer dumping.